flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java

```diff
@@ -0,0 +1,381 @@
+
+	/** The taskmanager pod name pattern is {clusterId}-{taskmanager}-{attemptId}-{podIndex}. */
+	private static final String TASK_MANAGER_POD_FORMAT = "%s-taskmanager-%d-%d";
+
+	private final ConcurrentMap<ResourceID, KubernetesWorkerNode> workerNodeMap;
+
+	private final int numberOfTaskSlots;
+
+	private final TaskExecutorResourceSpec taskExecutorResourceSpec;
+
+	private final int defaultTaskManagerMemoryMB;
+
+	private final double defaultCpus;
+
+	private final Collection<ResourceProfile> slotsPerWorker;
+
+	private final Configuration flinkConfig;
+
+	/** Flink configuration uploaded by client. */
+	private final Configuration flinkClientConfig;
+
+	/** When ResourceManager failover, the max attempt should recover. */
+	private final AtomicLong currentMaxAttemptId = new AtomicLong(0);
+
+	private final AtomicLong currentMaxPodId = new AtomicLong(0);
+
+	private final String clusterId;
+
+	private FlinkKubeClient kubeClient;
+
+	/** The number of pods requested, but not yet granted. */
+	private int numPendingPodRequests;
+
+	public KubernetesResourceManager(
+		RpcService rpcService,
+		String resourceManagerEndpointId,
+		ResourceID resourceId,
+		Configuration flinkConfig,
+		HighAvailabilityServices highAvailabilityServices,
+		HeartbeatServices heartbeatServices,
+		SlotManager slotManager,
+		JobLeaderIdService jobLeaderIdService,
+		ClusterInformation clusterInformation,
+		FatalErrorHandler fatalErrorHandler,
+		ResourceManagerMetricGroup resourceManagerMetricGroup) {
+		super(
+			rpcService,
+			resourceManagerEndpointId,
+			resourceId,
+			highAvailabilityServices,
+			heartbeatServices,
+			slotManager,
+			jobLeaderIdService,
+			clusterInformation,
+			fatalErrorHandler,
+			resourceManagerMetricGroup);
+
+		this.flinkConfig = flinkConfig;
+		this.clusterId = flinkConfig.getString(KubernetesConfigOptions.CLUSTER_ID);
+
+		this.workerNodeMap = new ConcurrentHashMap<>();
+
+		this.numPendingPodRequests = 0;
+
+		this.numberOfTaskSlots = flinkConfig.getInteger(TaskManagerOptions.NUM_TASK_SLOTS);
+		this.taskExecutorResourceSpec = TaskExecutorResourceUtils.resourceSpecFromConfig(flinkConfig);
+		this.defaultTaskManagerMemoryMB = taskExecutorResourceSpec.getTotalProcessMemorySize().getMebiBytes();
+		this.defaultCpus = flinkConfig.getDouble(KubernetesConfigOptions.TASK_MANAGER_CPU, numberOfTaskSlots);
+
+		this.slotsPerWorker = createWorkerSlotProfiles(flinkConfig);
+
+		// Load the flink config uploaded by flink client
+		this.flinkClientConfig = GlobalConfiguration.loadConfiguration();
+	}
+
+	@Override
+	protected void initialize() throws ResourceManagerException {
+		LOG.info("Initializing Kubernetes client.");
+		LOG.info("KubernetesResourceManager.initialize clusterId:{}", clusterId);
+
+		this.kubeClient = createFlinkKubeClient();
+
+		try {
+			getWorkerNodesFromPreviousAttempts();
+		} catch (Exception e) {
+			throw new ResourceManagerException(e);
```

Comment:
If we make `KubernetesWorkerNode#getAttempt()` throw `ResourceManagerException` directly instead of current `Execption`, then we can avoid wrapping this exception here again and remove `try..catch` for simple.

Reply:
Make sense. I will fix.